client/src/main/java/io/pravega/client/segment/impl/SegmentInputStreamImpl.java

```diff
@@ -181,7 +181,7 @@ private void verifyIsAtCorrectOffset(WireCommands.SegmentRead segmentRead) {
      * Issues a request if there is enough room for another request, and we aren't already waiting on one
      */
     private void issueRequestIfNeeded() {
-        if (!receivedEndOfSegment && buffer.capacityAvailable() > readLength) {
+        if (!receivedEndOfSegment && buffer.capacityAvailable() >= readLength) {
```

Comment:
This line is good but it will not prevent the NPE.
we need to modify the loop on line 147 to look like:
```
        while (result.hasRemaining()) {
            issueRequestIfNeeded();  // <- added
            handleRequest();
            offset += buffer.read(result);
        }
```

Reply:
You might be right, but I'm not entirely convinced and let me give you my argument so that you can show me  where it is incorrect. The flow I observed is the following:

1. We call `SegmentInputStreamImpl.read`, which calls `SegmentInputStreamImpl.readEventData`.
2. It invokes `SegmentInputStreamImpl.handleRequest` once in line 131.
3. It executes the `if` block of line 167, which sets `outstandingRequest` to `null` and invokes `SegmentInputStreamImpl.issueRequestIfNeeded`. The expectation as I understand this flow is that the call to `issueRequestIfNeeded` is that it will populate `outstadingRequest` because it is null. However, it doesn't because of the sign mismatch, which is fixed by making it greater or equal.

Adding `issueRequestIfNeeded` in line 147 is not wrong, but it does not seem to match my understanding of the flow because `outstandingRequest` should really have been populated previously in the execution of the while loop of line 124 (I'm assuming this will run for at least one interaction, is that correct?). The only way I have to justify the addition of the line you suggest is by assuming that it is possible for the while loop of line 124 to not run for a single iteration during an invocation of `readNextEvent`.

I have actually implemented your suggestion for the test case, and I couldn't observe the NPE. In fact, you can see that the latest Travis build has been successful and it contains the test case changes, but not the addition of `issueRequestIfNeeded();` to line 147.

Let me know what you think and help me understand the expected flow here better, please.

Reply:
Unfortunately the test case being an integration test, the timing to repro the more complicated case is not guaranteed. If you run the test case in a debugger and set breakpoints on the first read call and then do 'step over' three times to introduce a pause between the reads to allow other threads to complete the future reads, the test will fail.

The reason for the issue is what you highlight, the loop on line 124 may never execute. It's condition is there is less than 8 bytes in the buffer. Which under normal circumstances is not the case. 
So what ends up happening is that fillBuffer() on 123 moves the data from the outstanding request into the buffer, and then calls issueRequestIfNeeded(), but it does not issue a request because the buffer is more than half full. Then on 145 when the data is read out of the buffer it may be that the event being read is larger than the buffer so it enters the while loop on 146 and gets the NPE. If there were an issueRequestIfNeeded() call between the read on 145 and the handleRequest() on 147 then it would not throw a null pointer.

The case you fixed is just a special case of this error condition where the buffer is empty before readEventData() is called. In this case when fillBuffer() invokes issueRequestIfNeeded() and it will end up issuing a request because the pending request fills exactly half of the buffer. But if there were a few extra bytes left in the buffer before readEventData(), then more than half would be filled and no request would be issued.

Reply:
I believe that the `issueRequestIfNeeded()` could also go right before the loop. But if we wanted to avoid calling it in the event that there is not data remaining we would have to do:
```
if(result.hasRemaining()) {
    issueRequestIfNeeded();
    while (result.hasRemaining()) {
        handleRequest();
        offset += buffer.read(result);
    }
}
```
Which is a bit longer...

Reply:
Added call to `issueRequestIfNeeded`.